File: level1/src/main/java/org/com/level/Manchester.java
package org.com.level;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.Locale;
import java.util.Map;
import java.util.Scanner;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Collector;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class Manchester {
    public static final Map<String, Comparator<Player>> COMPARATOR = Map.of(
            "goals", Comparator.comparingInt(Player :: goals),
            "speed", Comparator.comparingInt(Player :: speed),
            "assists", Comparator.comparingInt(Player :: assists),
            "passing", Comparator.comparingInt(Player :: passingAccuracy),
            "defensivePlayer",
            Comparator.comparingInt(Player :: passingAccuracy)
    );
    private static Scanner scanner = new Scanner(System.in);
    private static Map<Integer, Player> manchester = Map.of(
            8, new Player("Bruno Fernandes", 5, 6, 9, 10, 3, 8),
            11, new Player("Rasmus Hojlund", 12, 8, 2, 6, 2, 11),
            5, new Player("Harry Maguire", 1, 5, 1, 7, 9, 5),
            1, new Player("Alejandro Garnacho", 8, 7, 8, 6, 0, 17),
            7, new Player("Mason Mount", 2, 6, 14, 8, 1, 7)
    );

    public static void main(String[] args) {
        var shouldContinue = true;
        while (shouldContinue) {
            printMenu();
            switch (scanner.nextInt()) {
                case 1 -> {
                    System.out.println("give the jersey number");
                    System.out.println(manchester.get(scanner.nextInt()));
                }
                case 2 -> {
                    requireData result = getRequireData();
                    var playerMap = getComparationMap(
                            result.statistics(),
                            result.player1(),
                            result.player2(),
                            COMPARATOR);
                    printComparation(playerMap);
                }
                case 3 -> printTopPlayerByStatistic("speed");
                case 4 -> printTopPlayerByStatistic("goals");
                case 5 -> printTopPlayerByStatistic("assists");
                case 6 -> printTopPlayerByStatistic("passing");
                case 7 -> printTopPlayerByStatistic("defensivePlayer");
                case 9 -> shouldContinue = false;
            }
        }
    }

    private static void printTopPlayerByStatistic(String st) {
        System.out.println(manchester.values().stream().max(COMPARATOR.get(st)).get());
    }

    private static requireData getRequireData() {
        var statistics = new ArrayList<String>();
        System.out.println("give the jersey number of player one");
        var player1 = manchester.get(scanner.nextInt());
        System.out.println("give the jersey number of player two");
        var player2 = manchester.get(scanner.nextInt());

        var shouldAksStatistic = true;
        while (shouldAksStatistic) {
            System.out.println("Select player's statistic to compare. 'Q' to stop");
            String next = scanner.next().toLowerCase(Locale.ROOT);
            shouldAksStatistic = (! next.equals("q")) && statistics.add(next);
        }
        return new requireData(statistics, player1, player2);
    }

    private record requireData(
            ArrayList<String> statistics, Player player1, Player player2
    ) {}

    private static Map<String, Player> getComparationMap(ArrayList<String> statistics,
                                                         Player player1, Player player2
            , Map<String, Comparator<Player>> comparator) {
        return statistics
                .stream()
                .map(comparatorFromStatisticToStaAndPlayer(player1, player2, comparator))
                .collect(makeAMap());
    }

    private static void printComparation(Map<String, Player> list) {
        list.forEach((key, player) -> System.out.println("Statistic: " + key + " Player" +
                                                         ":" + player.name()));
    }

    private static Collector<Map.Entry<String, Player>, ?, Map<String, Player>> makeAMap() {
        return Collectors.toMap(Map.Entry :: getKey, Map.Entry :: getValue);
    }

    private static Function<String, Map.Entry<String, Player>> comparatorFromStatisticToStaAndPlayer(Player player1, Player player2, Map<String, Comparator<Player>> comparator) {
        return st -> Map.entry(st,
                Stream.of(player1, player2).max(comparator.get(st)).get());
    }

    private static void printMenu() {
        System.out.println("1) Show player characteristics by Jersey Number");
        System.out.println("2) Compare two players");
        System.out.println("3) Show the fastest player");
        System.out.println("4) Show the top goal scorer");
        System.out.println("5) Show the player with most assist");
        System.out.println("6) Show the highest passing accuracy");
        System.out.println("7) Show the better defense player");
        System.out.println("8) to exit...!");
    }

    public static Player searchBy(String characteristic, int comparator) {
        Map<String, Predicate<Player>> playerMap = Map.of(
                "goals", player -> player.goals().equals(comparator),
                "speed", player -> player.speed().equals(comparator),
                "assists", player -> player.assists().equals(comparator),
                "accurate", player -> player.passingAccuracy().equals(comparator),
                "defensivePlayer", player -> player.defensive().equals(comparator)
        );
        return manchester
                .values()
                .stream()
                .filter(playerMap.get(characteristic))
                .findFirst()
                .orElseGet(() -> new Player("", 0, 0, 0, 0, 0, 0));
    }
}

File: manchester.py
'''Manchester United FC has hired you as a developer. Develop a program that helps the coach identify their 
fastest player, player with the most goals, assists, passing accuracy, and defensive involvements.
The system should also allow comparison between two players. Use the following player profiles:

Bruno Fernandes: 5 goals, 6 points in speed, 9 points in assists, 10 points in passing accuracy, 3 defensive 
involvements. Corresponds to jersey number 8.
Rasmus Hojlund: 12 goals, 8 points in speed, 2 points in assists, 6 points in passing accuracy, 2 defensive 
involvements. Corresponds to jersey number 11.
Harry Maguire: 1 goal, 5 points in speed, 1 point in assists, 7 points in passing accuracy, 9 defensive 
involvements. Corresponds to jersey number 5.
Alejandro Garnacho: 8 goals, 7 points in speed, 8 points in assists, 6 points in passing accuracy, 0 defensive
 involvements. Corresponds to jersey number 17.
Mason Mount: 2 goals, 6 points in speed, 4 points in assists, 8 points in passing accuracy, 1 defensive 
involvement. Corresponds to jersey number 7.
The program functions as follows: The coach accesses the system and encounters a menu with the following 
options:
Player Review: By entering the player's jersey number, they can access the player's characteristics.
Compare two players: The system prompts for two jersey numbers and displays the data of both players on screen.
Identify the fastest player: Displays the player with the most points in speed.
Identify the top goal scorer: Displays the player with the most points in goals.
Identify the player with the most assists: Displays the player with the most points in assists.
Identify the player with the highest passing accuracy: Displays the player with the most points in passing 
accuracy.
Identify the player with the most defensive involvements: Displays the player with the most points in 
defensive involvements.
The system should also allow returning to the main menu.
'''

class Player:
    
    def __init__(self,name:str,goals:int, speed:int, assists:int, passaccuracy:int, defense:int) -> None:
        self.name=name
        self.goals=goals
        self.speed=speed
        self.assists=assists
        self.passaccuracy=passaccuracy
        self.defense=defense

    def __repr__(self) -> str:
        return f"name: {self.name}\n goals: {self.goals}\nspeed: {self.speed}\nassists: {self.assists}\npass accuracy: {self.passaccuracy}\ndefense: {self.defense}\n"
            

players = {
        8:Player("Bruno Fernandes", 5, 6, 9, 10, 3), 11:Player("Rasmus Hojlund", 12, 8, 2, 6, 2),
        5:Player("Harry Maguire", 1, 5, 1, 7, 9), 17:Player("Alejandro Garnacho", 8, 7, 8, 6, 0), 
        7:Player("Mason Mount", 2, 6, 4, 8, 1)
    }
def search_player(number:int)->bool:
    return True if number in players.keys() else False
def player_review(number:int)->Player:
    return players[number] if search_player(number) else "The player isn't listed or doesn't exist"

def comparison(player1:int,player2:int)->Player:
    if search_player(player1) and search_player(player2):return players[player1], players[player2]
    else: return print("The player isn't listed or doesn't exist")
    
def top_player_by(stat:str):
    if stat in ["goals", "speed", "assists", "passaccuracy", "defense"]:
        return max(players.values(), key=lambda item: getattr(item, stat))
    else: 
        print("Try again, maybe you had a Typo.")
        return top_player_by(stat)
    
def main():
    choice = int(input("choose between the following options:\n1. Player review.\n2. Compare two players of your choice.\n3. The top player by a stat of your choice.\n4. Exit.\n"))
    control = "y"
    def return_to_menu(control):
        control=input("do you want to return to the main menu? y/n")
        if control =="y":
                return main()
        else: print("Thank you for using the player system. Good bye!")

    match(choice):
        case 1:
            number = int(input("Type the player's number."))
            print(player_review(number))
            return_to_menu(control)
            
        case 2:
            player1= int(input("Type the first player's number."))
            player2= int(input("Type the second player's number."))
            print(comparison(player1,player2))
            return_to_menu(control)
        case 3:
            stat=input("Type the stat you want to review:\ngoals, speed, assists, passaccuracy, defense")
            print(top_player_by(stat))
            return_to_menu(control)
        case 4:
            print("Thank you for using the player system. Good bye!")
        case _: 
            print("incorrect option.")
            main()
if __name__ == "__main__":
    main()